{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CONFIGURATION"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "pd.set_option('display.max_columns', 500)\n",
    "pd.set_option('display.width', 200)\n",
    "import numpy as np\n",
    "import copy\n",
    "import sys\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Hartwig imports\n",
    "sys.path.append(\"/Users/peterpriestley/hmf/repos/scripts/analysisscripts\") \n",
    "import analyseVCF as aVCF\n",
    "import venn as vn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#VCF\n",
    "VCF_PATH = \"/Users/peterpriestley/hmf/analyses/ensembleRuleTesting/\"\n",
    "\n",
    "# BED\n",
    "#BED_PATH = \"/Users/peterpriestley/hmf/analyses/giabTruthsets/\"\n",
    "#BED_FILE_NAME = \"NA12878_GIAB_highconf_IllFB-IllGATKHC-CG-Ion-Solid_ALLCHROM_v3.2.2_highconf.bed\"\n",
    "BED_PATH = \"/Users/peterpriestley/hmf/analyses/slices/\"\n",
    "#BED_FILE_NAME = \"cosmicCancerCensus.bed\" \n",
    "BED_FILE_NAME = \"CPCT_Slicing.bed\"\n",
    "USE_BED = True\n",
    "\n",
    "# HEALTH CHECKS\n",
    "HEALTH_CSV_PATH=\"/Users/peterpriestley/hmf/analyses/healthCheckAnalysis/\"\n",
    "HEALTH_CSV_FILENAME=\"healthCheckFile.csv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# SPECIFY SAMPLES\n",
    "VCF_SAMPLES=[]\n",
    "for x in os.listdir(VCF_PATH):\n",
    "    if x[-10:] == \"sliced.vcf\":\n",
    "        VCF_SAMPLES.append(x)\n",
    "#VCF_SAMPLES = ['160903_HMFregCPCT_FR10302737_FR10304566_CPCT02070012.filtered_variants_snpEff_snpSift_Cosmicv76_GoNLv5_sliced.vcf']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3> LOAD VCF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "reading BED file\n",
      "Bed File Loaded\n",
      "reading vcf file: 160903_HMFregCPCT_FR10302737_FR10304566_CPCT02070012.filtered_variants_snpEff_snpSift_Cosmicv76_GoNLv5_sliced.vcf\n",
      "Number variants loaded: 67\n",
      "reading BED file\n",
      "Bed File Loaded\n",
      "reading vcf file: 160903_HMFregCPCT_FR10302740_FR10304570_CPCT02050049.filtered_variants_snpEff_snpSift_Cosmicv76_GoNLv5_sliced.vcf\n",
      "Number variants loaded: 65\n",
      "reading BED file\n",
      "Bed File Loaded\n",
      "reading vcf file: 160905_HMFregCPCT_FR10301943_FR12251966_CPCT02160001.filtered_variants_snpEff_snpSift_Cosmicv76_GoNLv5_sliced.vcf\n",
      "Number variants loaded: 73\n",
      "reading BED file\n",
      "Bed File Loaded\n",
      "reading vcf file: 160905_HMFregCPCT_FR10302743_FR10304571_CPCT02050050.filtered_variants_snpEff_snpSift_Cosmicv76_GoNLv5_sliced.vcf\n",
      "Number variants loaded: 57\n",
      "reading BED file\n",
      "Bed File Loaded\n",
      "reading vcf file: 160905_HMFregCPCT_FR12244549_FR12251919_CPCT02160002.filtered_variants_snpEff_snpSift_Cosmicv76_GoNLv5_sliced.vcf\n",
      "Number variants loaded: 72\n",
      "reading BED file\n",
      "Bed File Loaded\n",
      "reading vcf file: 160905_HMFregCPCT_FR12244559_FR12251744_CPCT02160003.filtered_variants_snpEff_snpSift_Cosmicv76_GoNLv5_sliced.vcf\n",
      "Number variants loaded: 71\n",
      "reading BED file\n",
      "Bed File Loaded\n",
      "reading vcf file: 160905_HMFregCPCT_FR12251688_FR10304573_CPCT02020348.filtered_variants_snpEff_snpSift_Cosmicv76_GoNLv5_sliced.vcf\n",
      "Number variants loaded: 63\n",
      "reading BED file\n",
      "Bed File Loaded\n",
      "reading vcf file: 160905_HMFregCPCT_FR12254135_FR10304574_CPCT02020350.filtered_variants_snpEff_snpSift_Cosmicv76_GoNLv5_sliced.vcf\n",
      "Number variants loaded: 69\n",
      "reading BED file\n",
      "Bed File Loaded\n",
      "reading vcf file: 160905_HMFregCPCT_FR12254668_FR12251934_CPCT02160004.filtered_variants_snpEff_snpSift_Cosmicv76_GoNLv5_sliced.vcf\n",
      "Number variants loaded: 66\n",
      "reading BED file\n",
      "Bed File Loaded\n",
      "reading vcf file: 160906_HMFregCPCT_FR10302726_FR10304569_CPCT02050048.filtered_variants_snpEff_snpSift_Cosmicv76_GoNLv5_sliced.vcf\n",
      "Number variants loaded: 62\n",
      "reading BED file\n",
      "Bed File Loaded\n",
      "reading vcf file: 160906_HMFregCPCT_FR12254247_FR12251653_CPCT02140013.filtered_variants_snpEff_snpSift_Cosmicv76_GoNLv5_sliced.vcf\n",
      "Number variants loaded: 65\n",
      "reading BED file\n",
      "Bed File Loaded\n",
      "reading vcf file: 160907_HMFregCPCT_FR10301830_FR12254137_CPCT02010325.filtered_variants_snpEff_snpSift_Cosmicv76_GoNLv5_sliced.vcf\n",
      "Number variants loaded: 54\n",
      "reading BED file\n",
      "Bed File Loaded\n",
      "reading vcf file: 160912_HMFregCPCT_FR03631928_FR12251853_CPCT02060014.filtered_variants_snpEff_snpSift_Cosmicv76_GoNLv5_sliced.vcf\n",
      "Number variants loaded: 61\n",
      "reading BED file\n",
      "Bed File Loaded\n",
      "reading vcf file: 160912_HMFregCPCT_FR03631934_FR12251918_CPCT02060015.filtered_variants_snpEff_snpSift_Cosmicv76_GoNLv5_sliced.vcf\n",
      "Number variants loaded: 59\n",
      "reading BED file\n",
      "Bed File Loaded\n",
      "reading vcf file: 160912_HMFregCPCT_FR10302565_FR12251890_CPCT02060018.filtered_variants_snpEff_snpSift_Cosmicv76_GoNLv5_sliced.vcf\n",
      "Number variants loaded: 73\n",
      "reading BED file\n",
      "Bed File Loaded\n",
      "reading vcf file: 160912_HMFregCPCT_FR10302729_FR12251900_CPCT02010393.filtered_variants_snpEff_snpSift_Cosmicv76_GoNLv5_sliced.vcf\n",
      "Number variants loaded: 70\n",
      "reading BED file\n",
      "Bed File Loaded\n",
      "reading vcf file: 160912_HMFregCPCT_FR10302734_FR12251682_CPCT02010394.filtered_variants_snpEff_snpSift_Cosmicv76_GoNLv5_sliced.vcf\n",
      "Number variants loaded: 66\n",
      "reading BED file\n",
      "Bed File Loaded\n",
      "reading vcf file: 160912_HMFregCPCT_FR10302735_FR12254182_CPCT02010396.filtered_variants_snpEff_snpSift_Cosmicv76_GoNLv5_sliced.vcf\n",
      "Number variants loaded: 59\n",
      "reading BED file\n",
      "Bed File Loaded\n",
      "reading vcf file: 160912_HMFregCPCT_FR10302736_FR12254270_CPCT02010390.filtered_variants_snpEff_snpSift_Cosmicv76_GoNLv5_sliced.vcf\n",
      "Number variants loaded: 52\n",
      "reading BED file\n",
      "Bed File Loaded\n",
      "reading vcf file: 160912_HMFregCPCT_FR12244669_FR12251782_CPCT02060013.filtered_variants_snpEff_snpSift_Cosmicv76_GoNLv5_sliced.vcf\n",
      "Number variants loaded: 56\n",
      "reading BED file\n",
      "Bed File Loaded\n",
      "reading vcf file: 160913_HMFregCPCT_FR03631938_FR10304593_CPCT02010375.filtered_variants_snpEff_snpSift_Cosmicv76_GoNLv5_sliced.vcf\n",
      "Number variants loaded: 59\n",
      "reading BED file\n",
      "Bed File Loaded\n",
      "reading vcf file: 160913_HMFregCPCT_FR10302013_FR12254173_CPCT02020293.filtered_variants_snpEff_snpSift_Cosmicv76_GoNLv5_sliced.vcf\n",
      "Number variants loaded: 72\n",
      "reading BED file\n",
      "Bed File Loaded\n",
      "reading vcf file: 160913_HMFregCPCT_FR10302559_FR12251995_CPCT02060016.filtered_variants_snpEff_snpSift_Cosmicv76_GoNLv5_sliced.vcf\n",
      "Number variants loaded: 73\n",
      "reading BED file\n",
      "Bed File Loaded\n",
      "reading vcf file: 160913_HMFregCPCT_FR10302733_FR12251521_CPCT02060022.filtered_variants_snpEff_snpSift_Cosmicv76_GoNLv5_sliced.vcf\n",
      "Number variants loaded: 65\n",
      "reading BED file\n",
      "Bed File Loaded\n",
      "reading vcf file: 160913_HMFregCPCT_FR10302746_FR12251942_CPCT02140014.filtered_variants_snpEff_snpSift_Cosmicv76_GoNLv5_sliced.vcf\n",
      "Number variants loaded: 71\n",
      "reading BED file\n",
      "Bed File Loaded\n",
      "reading vcf file: 160913_HMFregCPCT_FR12244671_FR12251958_CPCT02020334.filtered_variants_snpEff_snpSift_Cosmicv76_GoNLv5_sliced.vcf\n",
      "Number variants loaded: 63\n",
      "reading BED file\n",
      "Bed File Loaded\n",
      "reading vcf file: 160913_HMFregCPCT_FR12251622_FR12251894_CPCT02060019.filtered_variants_snpEff_snpSift_Cosmicv76_GoNLv5_sliced.vcf\n",
      "Number variants loaded: 81\n",
      "reading BED file\n",
      "Bed File Loaded\n",
      "reading vcf file: 160916_HMFregCPCT_FR10302712_FR12254167_CPCT02030261.filtered_variants_snpEff_snpSift_Cosmicv76_GoNLv5_sliced.vcf\n",
      "Number variants loaded: 68\n",
      "reading BED file\n",
      "Bed File Loaded\n",
      "reading vcf file: 160916_HMFregCPCT_FR10302739_FR12251877_CPCT02040035.filtered_variants_snpEff_snpSift_Cosmicv76_GoNLv5_sliced.vcf\n",
      "Number variants loaded: 66\n",
      "reading BED file\n",
      "Bed File Loaded\n",
      "reading vcf file: 160916_HMFregCPCT_FR10302747_FR12251536_CPCT02010398.filtered_variants_snpEff_snpSift_Cosmicv76_GoNLv5_sliced.vcf\n",
      "Number variants loaded: 66\n",
      "reading BED file\n",
      "Bed File Loaded\n",
      "reading vcf file: 160916_HMFregCPCT_FR10302748_FR12251953_CPCT02020358.filtered_variants_snpEff_snpSift_Cosmicv76_GoNLv5_sliced.vcf\n",
      "Number variants loaded: 72\n",
      "reading BED file\n",
      "Bed File Loaded\n",
      "reading vcf file: 160916_HMFregCPCT_FR10302749_FR12251983_CPCT02020357.filtered_variants_snpEff_snpSift_Cosmicv76_GoNLv5_sliced.vcf\n",
      "Number variants loaded: 60\n",
      "reading BED file\n",
      "Bed File Loaded\n",
      "reading vcf file: 160916_HMFregCPCT_FR10302780_FR12254138_CPCT02010397.filtered_variants_snpEff_snpSift_Cosmicv76_GoNLv5_sliced.vcf\n",
      "Number variants loaded: 75\n",
      "reading BED file\n",
      "Bed File Loaded\n",
      "reading vcf file: 160916_HMFregCPCT_FR10302781_FR12251599_CPCT02110006.filtered_variants_snpEff_snpSift_Cosmicv76_GoNLv5_sliced.vcf\n",
      "Number variants loaded: 65\n",
      "reading BED file\n",
      "Bed File Loaded\n",
      "reading vcf file: 160916_HMFregCPCT_FR12251560_FR12251906_CPCT02020351.filtered_variants_snpEff_snpSift_Cosmicv76_GoNLv5_sliced.vcf\n",
      "Number variants loaded: 62\n",
      "reading BED file\n",
      "Bed File Loaded\n",
      "reading vcf file: 160916_HMFregCPCT_FR12251705_FR12251527_CPCT02110007.filtered_variants_snpEff_snpSift_Cosmicv76_GoNLv5_sliced.vcf\n",
      "Number variants loaded: 57\n",
      "reading BED file\n",
      "Bed File Loaded\n",
      "reading vcf file: 160916_HMFregCPCT_FR12254273_FR12251547_CPCT02160005.filtered_variants_snpEff_snpSift_Cosmicv76_GoNLv5_sliced.vcf\n",
      "Number variants loaded: 56\n",
      "reading BED file\n",
      "Bed File Loaded\n",
      "reading vcf file: 160922_HMFregCPCT_FR10302720_FR03631828_CPCT02100026.filtered_variants_snpEff_snpSift_Cosmicv76_GoNLv5_sliced.vcf\n",
      "Number variants loaded: 65\n",
      "reading BED file\n",
      "Bed File Loaded\n",
      "reading vcf file: 160922_HMFregCPCT_FR10302744_FR03631829_CPCT02100027.filtered_variants_snpEff_snpSift_Cosmicv76_GoNLv5_sliced.vcf\n",
      "Number variants loaded: 67\n",
      "reading BED file\n",
      "Bed File Loaded\n",
      "reading vcf file: 160922_HMFregCPCT_FR10302751_FR12251992_CPCT02140015.filtered_variants_snpEff_snpSift_Cosmicv76_GoNLv5_sliced.vcf\n",
      "Number variants loaded: 65\n",
      "reading BED file\n",
      "Bed File Loaded\n",
      "reading vcf file: 160922_HMFregCPCT_FR10302779_FR12251970_CPCT02020360.filtered_variants_snpEff_snpSift_Cosmicv76_GoNLv5_sliced.vcf\n",
      "Number variants loaded: 68\n",
      "reading BED file\n",
      "Bed File Loaded\n",
      "reading vcf file: 160922_HMFregCPCT_FR10302782_FR12251860_CPCT02060023.filtered_variants_snpEff_snpSift_Cosmicv76_GoNLv5_sliced.vcf\n",
      "Number variants loaded: 71\n"
     ]
    }
   ],
   "source": [
    "#Load all VCF_SAMPLES to DF\n",
    "df = pd.DataFrame({})\n",
    "for VCF_SAMPLE in VCF_SAMPLES:\n",
    "    VCF_FILE_NAME = VCF_SAMPLE\n",
    "    SAMPLE_NAMES = {VCF_SAMPLE[40:52]+'R':'normal',VCF_SAMPLE[40:52]+'T':'tumor'}\n",
    "    if USE_BED:\n",
    "        bed = aVCF.loadBEDFile(BED_PATH,BED_FILE_NAME)\n",
    "        df = pd.concat([df,aVCF.loadVariantsFromVCF(VCF_PATH,VCF_FILE_NAME,SAMPLE_NAMES,VCF_SAMPLE[40:52],True,True,bed,False)])\n",
    "    else:\n",
    "        df = pd.concat([df,aVCF.loadVariantsFromVCF(VCF_PATH,VCF_FILE_NAME,SAMPLE_NAMES,VCF_SAMPLE[40:52],True)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#LOAD HEALTH_CHECKS\n",
    "hc = pd.read_csv(HEALTH_CSV_PATH + HEALTH_CSV_FILENAME)\n",
    "hc = hc[['SAMPLE','TUMOR_PERCENTAGE','TUMOR_TYPE','PIPELINE_VERSION']]\n",
    "hc = hc[hc['SAMPLE'].str[12]==\"T\"]   # only tumor rows\n",
    "hc['SAMPLE']=hc['SAMPLE'].str[:12]\n",
    "hc = hc.set_index(['SAMPLE'])\n",
    "df = pd.merge(df,hc,how='left', left_on='patientName',right_index=True)\n",
    "df.fillna(\"Unknown\", inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3> High Level Stats"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### LOH variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>consensus</th>\n",
       "      <th>False</th>\n",
       "      <th>True</th>\n",
       "      <th>All</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chromPos</th>\n",
       "      <th>patientName</th>\n",
       "      <th>annGene</th>\n",
       "      <th>normalSGT</th>\n",
       "      <th>tumorSGT</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">10:89720633</th>\n",
       "      <th>CPCT02040035</th>\n",
       "      <th>PTEN</th>\n",
       "      <th>0/1</th>\n",
       "      <th></th>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CPCT02050048</th>\n",
       "      <th>PTEN</th>\n",
       "      <th>0/1</th>\n",
       "      <th></th>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CPCT02060018</th>\n",
       "      <th>PTEN</th>\n",
       "      <th>0/1</th>\n",
       "      <th></th>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11:534242</th>\n",
       "      <th>CPCT02010393</th>\n",
       "      <th>HRAS</th>\n",
       "      <th>0/1</th>\n",
       "      <th>1/1</th>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13:32914236</th>\n",
       "      <th>CPCT02020357</th>\n",
       "      <th>BRCA2</th>\n",
       "      <th>0/1</th>\n",
       "      <th></th>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14:105246325</th>\n",
       "      <th>CPCT02010393</th>\n",
       "      <th>AKT1</th>\n",
       "      <th>0/1</th>\n",
       "      <th>1/1</th>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14:105246407</th>\n",
       "      <th>CPCT02160003</th>\n",
       "      <th>AKT1</th>\n",
       "      <th>0/1</th>\n",
       "      <th>1/1</th>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14:105246565</th>\n",
       "      <th>CPCT02160003</th>\n",
       "      <th>AKT1</th>\n",
       "      <th>0/1</th>\n",
       "      <th></th>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14:105258892</th>\n",
       "      <th>CPCT02010393</th>\n",
       "      <th>AKT1</th>\n",
       "      <th>0/1</th>\n",
       "      <th></th>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17:37868189</th>\n",
       "      <th>CPCT02020293</th>\n",
       "      <th>ERBB2</th>\n",
       "      <th>0/1</th>\n",
       "      <th></th>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17:37884037</th>\n",
       "      <th>CPCT02010393</th>\n",
       "      <th>ERBB2</th>\n",
       "      <th>0/1</th>\n",
       "      <th></th>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17:41197709</th>\n",
       "      <th>CPCT02020293</th>\n",
       "      <th>BRCA1</th>\n",
       "      <th>0/1</th>\n",
       "      <th></th>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">17:41223094</th>\n",
       "      <th>CPCT02010393</th>\n",
       "      <th>BRCA1</th>\n",
       "      <th>0/1</th>\n",
       "      <th></th>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CPCT02160003</th>\n",
       "      <th>BRCA1</th>\n",
       "      <th>0/1</th>\n",
       "      <th></th>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17:41234470</th>\n",
       "      <th>CPCT02010393</th>\n",
       "      <th>BRCA1</th>\n",
       "      <th>0/1</th>\n",
       "      <th></th>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17:41244435</th>\n",
       "      <th>CPCT02010393</th>\n",
       "      <th>BRCA1</th>\n",
       "      <th>0/1</th>\n",
       "      <th></th>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17:41244936</th>\n",
       "      <th>CPCT02010393</th>\n",
       "      <th>BRCA1</th>\n",
       "      <th>0/1</th>\n",
       "      <th></th>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17:41245237</th>\n",
       "      <th>CPCT02010393</th>\n",
       "      <th>BRCA1</th>\n",
       "      <th>0/1</th>\n",
       "      <th></th>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17:41245466</th>\n",
       "      <th>CPCT02010393</th>\n",
       "      <th>BRCA1</th>\n",
       "      <th>0/1</th>\n",
       "      <th></th>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17:41246481</th>\n",
       "      <th>CPCT02010393</th>\n",
       "      <th>BRCA1</th>\n",
       "      <th>0/1</th>\n",
       "      <th>1/1</th>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17:7578645</th>\n",
       "      <th>CPCT02010393</th>\n",
       "      <th>TP53</th>\n",
       "      <th>0/1</th>\n",
       "      <th></th>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">2:212578379</th>\n",
       "      <th>CPCT02010394</th>\n",
       "      <th>ERBB4</th>\n",
       "      <th>0/1</th>\n",
       "      <th>1/1</th>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CPCT02140015</th>\n",
       "      <th>ERBB4</th>\n",
       "      <th>0/1</th>\n",
       "      <th>1/1</th>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3:10183444</th>\n",
       "      <th>CPCT02020293</th>\n",
       "      <th>VHL</th>\n",
       "      <th>0/1</th>\n",
       "      <th></th>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">4:1795557</th>\n",
       "      <th>CPCT02010325</th>\n",
       "      <th>FGFR3</th>\n",
       "      <th>1/1</th>\n",
       "      <th></th>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CPCT02110006</th>\n",
       "      <th>FGFR3</th>\n",
       "      <th>1/1</th>\n",
       "      <th></th>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CPCT02160002</th>\n",
       "      <th>FGFR3</th>\n",
       "      <th>1/1</th>\n",
       "      <th></th>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4:1806012</th>\n",
       "      <th>CPCT02060019</th>\n",
       "      <th>FGFR3</th>\n",
       "      <th>0/1</th>\n",
       "      <th></th>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4:1808004</th>\n",
       "      <th>CPCT02010375</th>\n",
       "      <th>FGFR3</th>\n",
       "      <th>0/1</th>\n",
       "      <th></th>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4:1808006</th>\n",
       "      <th>CPCT02010375</th>\n",
       "      <th>FGFR3</th>\n",
       "      <th>0/1</th>\n",
       "      <th></th>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4:1808010</th>\n",
       "      <th>CPCT02010375</th>\n",
       "      <th>FGFR3</th>\n",
       "      <th>0/1</th>\n",
       "      <th></th>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>All</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <td>29</td>\n",
       "      <td>2</td>\n",
       "      <td>31.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "consensus                                            False True   All\n",
       "chromPos     patientName  annGene normalSGT tumorSGT                 \n",
       "10:89720633  CPCT02040035 PTEN    0/1                    1        1.0\n",
       "             CPCT02050048 PTEN    0/1                    1        1.0\n",
       "             CPCT02060018 PTEN    0/1                    1        1.0\n",
       "11:534242    CPCT02010393 HRAS    0/1       1/1               1   1.0\n",
       "13:32914236  CPCT02020357 BRCA2   0/1                    1        1.0\n",
       "14:105246325 CPCT02010393 AKT1    0/1       1/1          1        1.0\n",
       "14:105246407 CPCT02160003 AKT1    0/1       1/1          1        1.0\n",
       "14:105246565 CPCT02160003 AKT1    0/1                    1        1.0\n",
       "14:105258892 CPCT02010393 AKT1    0/1                    1        1.0\n",
       "17:37868189  CPCT02020293 ERBB2   0/1                    1        1.0\n",
       "17:37884037  CPCT02010393 ERBB2   0/1                    1        1.0\n",
       "17:41197709  CPCT02020293 BRCA1   0/1                    1        1.0\n",
       "17:41223094  CPCT02010393 BRCA1   0/1                    1        1.0\n",
       "             CPCT02160003 BRCA1   0/1                    1        1.0\n",
       "17:41234470  CPCT02010393 BRCA1   0/1                    1        1.0\n",
       "17:41244435  CPCT02010393 BRCA1   0/1                    1        1.0\n",
       "17:41244936  CPCT02010393 BRCA1   0/1                    1        1.0\n",
       "17:41245237  CPCT02010393 BRCA1   0/1                    1        1.0\n",
       "17:41245466  CPCT02010393 BRCA1   0/1                    1        1.0\n",
       "17:41246481  CPCT02010393 BRCA1   0/1       1/1               1   1.0\n",
       "17:7578645   CPCT02010393 TP53    0/1                    1        1.0\n",
       "2:212578379  CPCT02010394 ERBB4   0/1       1/1          1        1.0\n",
       "             CPCT02140015 ERBB4   0/1       1/1          1        1.0\n",
       "3:10183444   CPCT02020293 VHL     0/1                    1        1.0\n",
       "4:1795557    CPCT02010325 FGFR3   1/1                    1        1.0\n",
       "             CPCT02110006 FGFR3   1/1                    1        1.0\n",
       "             CPCT02160002 FGFR3   1/1                    1        1.0\n",
       "4:1806012    CPCT02060019 FGFR3   0/1                    1        1.0\n",
       "4:1808004    CPCT02010375 FGFR3   0/1                    1        1.0\n",
       "4:1808006    CPCT02010375 FGFR3   0/1                    1        1.0\n",
       "4:1808010    CPCT02010375 FGFR3   0/1                    1        1.0\n",
       "All                                                     29    2  31.0"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.pivot_table(df[(df.normalSGT<>\"\")&(df.normalSGT<>df.tumorSGT)],\\\n",
    "               values='pos', index=['chromPos','patientName','annGene','normalSGT','tumorSGT'], columns=['consensus'], aggfunc='count',margins=True).fillna(\"\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### HIGH and MODERATE impact NON-missense"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>consensus</th>\n",
       "      <th>False</th>\n",
       "      <th>All</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chromPos</th>\n",
       "      <th>patientName</th>\n",
       "      <th>annGene</th>\n",
       "      <th>annWorstEffect</th>\n",
       "      <th>annWorstImpact</th>\n",
       "      <th>ref</th>\n",
       "      <th>normalallele</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10:89692784</th>\n",
       "      <th>CPCT02140013</th>\n",
       "      <th>PTEN</th>\n",
       "      <th>disruptive_inframe_insertion</th>\n",
       "      <th>MODERATE</th>\n",
       "      <th>T</th>\n",
       "      <th></th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10:89717731</th>\n",
       "      <th>CPCT02060015</th>\n",
       "      <th>PTEN</th>\n",
       "      <th>frameshift_variant</th>\n",
       "      <th>HIGH</th>\n",
       "      <th>T</th>\n",
       "      <th></th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10:89720741</th>\n",
       "      <th>CPCT02140015</th>\n",
       "      <th>PTEN</th>\n",
       "      <th>stop_gained</th>\n",
       "      <th>HIGH</th>\n",
       "      <th>C</th>\n",
       "      <th></th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10:89720798</th>\n",
       "      <th>CPCT02160001</th>\n",
       "      <th>PTEN</th>\n",
       "      <th>frameshift_variant</th>\n",
       "      <th>HIGH</th>\n",
       "      <th>GTACT</th>\n",
       "      <th></th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16:68856020</th>\n",
       "      <th>CPCT02160001</th>\n",
       "      <th>CDH1</th>\n",
       "      <th>stop_gained</th>\n",
       "      <th>HIGH</th>\n",
       "      <th>C</th>\n",
       "      <th></th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17:7573976</th>\n",
       "      <th>CPCT02060019</th>\n",
       "      <th>TP53</th>\n",
       "      <th>stop_gained</th>\n",
       "      <th>HIGH</th>\n",
       "      <th>T</th>\n",
       "      <th></th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17:7576851</th>\n",
       "      <th>CPCT02140015</th>\n",
       "      <th>TP53</th>\n",
       "      <th>splice_donor_variant&amp;intron_variant</th>\n",
       "      <th>HIGH</th>\n",
       "      <th>A</th>\n",
       "      <th></th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17:7576855</th>\n",
       "      <th>CPCT02010397</th>\n",
       "      <th>TP53</th>\n",
       "      <th>stop_gained&amp;splice_region_variant</th>\n",
       "      <th>HIGH</th>\n",
       "      <th>G</th>\n",
       "      <th></th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17:7577085</th>\n",
       "      <th>CPCT02030261</th>\n",
       "      <th>TP53</th>\n",
       "      <th>stop_gained</th>\n",
       "      <th>HIGH</th>\n",
       "      <th>C</th>\n",
       "      <th></th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17:7577552</th>\n",
       "      <th>CPCT02060014</th>\n",
       "      <th>TP53</th>\n",
       "      <th>frameshift_variant</th>\n",
       "      <th>HIGH</th>\n",
       "      <th>CA</th>\n",
       "      <th></th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17:7578212</th>\n",
       "      <th>CPCT02010393</th>\n",
       "      <th>TP53</th>\n",
       "      <th>stop_gained</th>\n",
       "      <th>HIGH</th>\n",
       "      <th>G</th>\n",
       "      <th></th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17:7578274</th>\n",
       "      <th>CPCT02010396</th>\n",
       "      <th>TP53</th>\n",
       "      <th>disruptive_inframe_deletion</th>\n",
       "      <th>MODERATE</th>\n",
       "      <th>TGAG</th>\n",
       "      <th></th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17:7578463</th>\n",
       "      <th>CPCT02160003</th>\n",
       "      <th>TP53</th>\n",
       "      <th>frameshift_variant</th>\n",
       "      <th>HIGH</th>\n",
       "      <th>CGGGTGCCGGGCGG</th>\n",
       "      <th></th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17:7579429</th>\n",
       "      <th>CPCT02060018</th>\n",
       "      <th>TP53</th>\n",
       "      <th>frameshift_variant</th>\n",
       "      <th>HIGH</th>\n",
       "      <th>TG</th>\n",
       "      <th></th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17:7579462</th>\n",
       "      <th>CPCT02020357</th>\n",
       "      <th>TP53</th>\n",
       "      <th>frameshift_variant</th>\n",
       "      <th>HIGH</th>\n",
       "      <th>AG</th>\n",
       "      <th></th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3:10183791</th>\n",
       "      <th>CPCT02110007</th>\n",
       "      <th>VHL</th>\n",
       "      <th>frameshift_variant</th>\n",
       "      <th>HIGH</th>\n",
       "      <th>TA</th>\n",
       "      <th></th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5:112173917</th>\n",
       "      <th>CPCT02050048</th>\n",
       "      <th>APC</th>\n",
       "      <th>stop_gained</th>\n",
       "      <th>HIGH</th>\n",
       "      <th>C</th>\n",
       "      <th></th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5:112175399</th>\n",
       "      <th>CPCT02110006</th>\n",
       "      <th>APC</th>\n",
       "      <th>stop_gained</th>\n",
       "      <th>HIGH</th>\n",
       "      <th>A</th>\n",
       "      <th></th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5:112176017</th>\n",
       "      <th>CPCT02060015</th>\n",
       "      <th>APC</th>\n",
       "      <th>stop_gained</th>\n",
       "      <th>HIGH</th>\n",
       "      <th>G</th>\n",
       "      <th></th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9:21971154</th>\n",
       "      <th>CPCT02010393</th>\n",
       "      <th>CDKN2A</th>\n",
       "      <th>frameshift_variant</th>\n",
       "      <th>HIGH</th>\n",
       "      <th>CG</th>\n",
       "      <th></th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>All</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <td>20.0</td>\n",
       "      <td>20.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "consensus                                                                                                        False   All\n",
       "chromPos    patientName  annGene annWorstEffect                      annWorstImpact ref            normalallele             \n",
       "10:89692784 CPCT02140013 PTEN    disruptive_inframe_insertion        MODERATE       T                              1.0   1.0\n",
       "10:89717731 CPCT02060015 PTEN    frameshift_variant                  HIGH           T                              1.0   1.0\n",
       "10:89720741 CPCT02140015 PTEN    stop_gained                         HIGH           C                              1.0   1.0\n",
       "10:89720798 CPCT02160001 PTEN    frameshift_variant                  HIGH           GTACT                          1.0   1.0\n",
       "16:68856020 CPCT02160001 CDH1    stop_gained                         HIGH           C                              1.0   1.0\n",
       "17:7573976  CPCT02060019 TP53    stop_gained                         HIGH           T                              1.0   1.0\n",
       "17:7576851  CPCT02140015 TP53    splice_donor_variant&intron_variant HIGH           A                              1.0   1.0\n",
       "17:7576855  CPCT02010397 TP53    stop_gained&splice_region_variant   HIGH           G                              1.0   1.0\n",
       "17:7577085  CPCT02030261 TP53    stop_gained                         HIGH           C                              1.0   1.0\n",
       "17:7577552  CPCT02060014 TP53    frameshift_variant                  HIGH           CA                             1.0   1.0\n",
       "17:7578212  CPCT02010393 TP53    stop_gained                         HIGH           G                              1.0   1.0\n",
       "17:7578274  CPCT02010396 TP53    disruptive_inframe_deletion         MODERATE       TGAG                           1.0   1.0\n",
       "17:7578463  CPCT02160003 TP53    frameshift_variant                  HIGH           CGGGTGCCGGGCGG                 1.0   1.0\n",
       "17:7579429  CPCT02060018 TP53    frameshift_variant                  HIGH           TG                             1.0   1.0\n",
       "17:7579462  CPCT02020357 TP53    frameshift_variant                  HIGH           AG                             1.0   1.0\n",
       "3:10183791  CPCT02110007 VHL     frameshift_variant                  HIGH           TA                             1.0   1.0\n",
       "5:112173917 CPCT02050048 APC     stop_gained                         HIGH           C                              1.0   1.0\n",
       "5:112175399 CPCT02110006 APC     stop_gained                         HIGH           A                              1.0   1.0\n",
       "5:112176017 CPCT02060015 APC     stop_gained                         HIGH           G                              1.0   1.0\n",
       "9:21971154  CPCT02010393 CDKN2A  frameshift_variant                  HIGH           CG                             1.0   1.0\n",
       "All                                                                                                               20.0  20.0"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.pivot_table(df[((df['annWorstImpact']=='MODERATE')|(df['annWorstImpact']=='HIGH'))&(~df.annAllEffects.str.contains(\"missense\"))&(df.normalSGT==\"\")],\\\n",
    "               values='pos', index=['chromPos','patientName','annGene','annWorstEffect','annWorstImpact','ref','normalallele'], columns=['consensus'], aggfunc='count',margins=True).fillna(\"\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Missense"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>consensus</th>\n",
       "      <th>False</th>\n",
       "      <th>All</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chromPos</th>\n",
       "      <th>patientName</th>\n",
       "      <th>annGene</th>\n",
       "      <th>normalSGT</th>\n",
       "      <th>tumorSGT</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10:89624305</th>\n",
       "      <th>CPCT02110007</th>\n",
       "      <th>PTEN</th>\n",
       "      <th></th>\n",
       "      <th>0/1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10:89720677</th>\n",
       "      <th>CPCT02060015</th>\n",
       "      <th>PTEN</th>\n",
       "      <th></th>\n",
       "      <th>0/1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12:25398284</th>\n",
       "      <th>CPCT02110006</th>\n",
       "      <th>KRAS</th>\n",
       "      <th></th>\n",
       "      <th>0/1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13:32910968</th>\n",
       "      <th>CPCT02030261</th>\n",
       "      <th>BRCA2</th>\n",
       "      <th></th>\n",
       "      <th>0/1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13:49033926</th>\n",
       "      <th>CPCT02020357</th>\n",
       "      <th>RB1</th>\n",
       "      <th></th>\n",
       "      <th>0/1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16:68842328</th>\n",
       "      <th>CPCT02050048</th>\n",
       "      <th>CDH1</th>\n",
       "      <th></th>\n",
       "      <th>0/1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17:37881000</th>\n",
       "      <th>CPCT02060019</th>\n",
       "      <th>ERBB2</th>\n",
       "      <th></th>\n",
       "      <th>0/1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17:7577536</th>\n",
       "      <th>CPCT02110006</th>\n",
       "      <th>TP53</th>\n",
       "      <th></th>\n",
       "      <th>0/1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17:7577538</th>\n",
       "      <th>CPCT02010390</th>\n",
       "      <th>TP53</th>\n",
       "      <th></th>\n",
       "      <th>0/1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17:7577539</th>\n",
       "      <th>CPCT02050048</th>\n",
       "      <th>TP53</th>\n",
       "      <th></th>\n",
       "      <th>0/1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17:7577566</th>\n",
       "      <th>CPCT02020293</th>\n",
       "      <th>TP53</th>\n",
       "      <th></th>\n",
       "      <th>0/1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17:7577580</th>\n",
       "      <th>CPCT02020334</th>\n",
       "      <th>TP53</th>\n",
       "      <th></th>\n",
       "      <th>0/1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17:7578268</th>\n",
       "      <th>CPCT02110007</th>\n",
       "      <th>TP53</th>\n",
       "      <th></th>\n",
       "      <th>0/1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17:7578393</th>\n",
       "      <th>CPCT02140015</th>\n",
       "      <th>TP53</th>\n",
       "      <th></th>\n",
       "      <th>0/1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17:7578454</th>\n",
       "      <th>CPCT02020348</th>\n",
       "      <th>TP53</th>\n",
       "      <th></th>\n",
       "      <th>0/1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17:7578523</th>\n",
       "      <th>CPCT02020351</th>\n",
       "      <th>TP53</th>\n",
       "      <th></th>\n",
       "      <th>0/1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17:7578528</th>\n",
       "      <th>CPCT02020350</th>\n",
       "      <th>TP53</th>\n",
       "      <th></th>\n",
       "      <th>0/1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1:115256529</th>\n",
       "      <th>CPCT02010393</th>\n",
       "      <th>NRAS</th>\n",
       "      <th></th>\n",
       "      <th>0/1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1:115256530</th>\n",
       "      <th>CPCT02020358</th>\n",
       "      <th>NRAS</th>\n",
       "      <th></th>\n",
       "      <th>0/1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2:16082244</th>\n",
       "      <th>CPCT02160001</th>\n",
       "      <th>MYCN</th>\n",
       "      <th></th>\n",
       "      <th>0/1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2:209113113</th>\n",
       "      <th>CPCT02020358</th>\n",
       "      <th>IDH1</th>\n",
       "      <th></th>\n",
       "      <th>0/1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3:178916930</th>\n",
       "      <th>CPCT02100027</th>\n",
       "      <th>PIK3CA</th>\n",
       "      <th></th>\n",
       "      <th>0/1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3:178916947</th>\n",
       "      <th>CPCT02140015</th>\n",
       "      <th>PIK3CA</th>\n",
       "      <th></th>\n",
       "      <th>0/1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3:178936082</th>\n",
       "      <th>CPCT02030261</th>\n",
       "      <th>PIK3CA</th>\n",
       "      <th></th>\n",
       "      <th>0/1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3:178936091</th>\n",
       "      <th>CPCT02040035</th>\n",
       "      <th>PIK3CA</th>\n",
       "      <th></th>\n",
       "      <th>0/1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3:178936095</th>\n",
       "      <th>CPCT02110007</th>\n",
       "      <th>PIK3CA</th>\n",
       "      <th></th>\n",
       "      <th>0/1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3:178952085</th>\n",
       "      <th>CPCT02030261</th>\n",
       "      <th>PIK3CA</th>\n",
       "      <th></th>\n",
       "      <th>0/1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4:55141084</th>\n",
       "      <th>CPCT02010325</th>\n",
       "      <th>PDGFRA</th>\n",
       "      <th></th>\n",
       "      <th>0/1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5:112175180</th>\n",
       "      <th>CPCT02140013</th>\n",
       "      <th>APC</th>\n",
       "      <th></th>\n",
       "      <th>0/1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6:152419923</th>\n",
       "      <th>CPCT02060022</th>\n",
       "      <th>ESR1</th>\n",
       "      <th></th>\n",
       "      <th>0/1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6:152419926</th>\n",
       "      <th>CPCT02160002</th>\n",
       "      <th>ESR1</th>\n",
       "      <th></th>\n",
       "      <th>0/1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">7:140453136</th>\n",
       "      <th>CPCT02010397</th>\n",
       "      <th>BRAF</th>\n",
       "      <th></th>\n",
       "      <th>0/1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CPCT02020293</th>\n",
       "      <th>BRAF</th>\n",
       "      <th></th>\n",
       "      <th>0/1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CPCT02020360</th>\n",
       "      <th>BRAF</th>\n",
       "      <th></th>\n",
       "      <th>0/1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CPCT02060015</th>\n",
       "      <th>BRAF</th>\n",
       "      <th></th>\n",
       "      <th>0/1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7:55223550</th>\n",
       "      <th>CPCT02010393</th>\n",
       "      <th>EGFR</th>\n",
       "      <th></th>\n",
       "      <th>0/1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8:128752740</th>\n",
       "      <th>CPCT02070012</th>\n",
       "      <th>MYC</th>\n",
       "      <th></th>\n",
       "      <th>0/1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9:21971107</th>\n",
       "      <th>CPCT02060019</th>\n",
       "      <th>CDKN2A</th>\n",
       "      <th></th>\n",
       "      <th>0/1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>All</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <td>38.0</td>\n",
       "      <td>38.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "consensus                                            False   All\n",
       "chromPos    patientName  annGene normalSGT tumorSGT             \n",
       "10:89624305 CPCT02110007 PTEN              0/1         1.0   1.0\n",
       "10:89720677 CPCT02060015 PTEN              0/1         1.0   1.0\n",
       "12:25398284 CPCT02110006 KRAS              0/1         1.0   1.0\n",
       "13:32910968 CPCT02030261 BRCA2             0/1         1.0   1.0\n",
       "13:49033926 CPCT02020357 RB1               0/1         1.0   1.0\n",
       "16:68842328 CPCT02050048 CDH1              0/1         1.0   1.0\n",
       "17:37881000 CPCT02060019 ERBB2             0/1         1.0   1.0\n",
       "17:7577536  CPCT02110006 TP53              0/1         1.0   1.0\n",
       "17:7577538  CPCT02010390 TP53              0/1         1.0   1.0\n",
       "17:7577539  CPCT02050048 TP53              0/1         1.0   1.0\n",
       "17:7577566  CPCT02020293 TP53              0/1         1.0   1.0\n",
       "17:7577580  CPCT02020334 TP53              0/1         1.0   1.0\n",
       "17:7578268  CPCT02110007 TP53              0/1         1.0   1.0\n",
       "17:7578393  CPCT02140015 TP53              0/1         1.0   1.0\n",
       "17:7578454  CPCT02020348 TP53              0/1         1.0   1.0\n",
       "17:7578523  CPCT02020351 TP53              0/1         1.0   1.0\n",
       "17:7578528  CPCT02020350 TP53              0/1         1.0   1.0\n",
       "1:115256529 CPCT02010393 NRAS              0/1         1.0   1.0\n",
       "1:115256530 CPCT02020358 NRAS              0/1         1.0   1.0\n",
       "2:16082244  CPCT02160001 MYCN              0/1         1.0   1.0\n",
       "2:209113113 CPCT02020358 IDH1              0/1         1.0   1.0\n",
       "3:178916930 CPCT02100027 PIK3CA            0/1         1.0   1.0\n",
       "3:178916947 CPCT02140015 PIK3CA            0/1         1.0   1.0\n",
       "3:178936082 CPCT02030261 PIK3CA            0/1         1.0   1.0\n",
       "3:178936091 CPCT02040035 PIK3CA            0/1         1.0   1.0\n",
       "3:178936095 CPCT02110007 PIK3CA            0/1         1.0   1.0\n",
       "3:178952085 CPCT02030261 PIK3CA            0/1         1.0   1.0\n",
       "4:55141084  CPCT02010325 PDGFRA            0/1         1.0   1.0\n",
       "5:112175180 CPCT02140013 APC               0/1         1.0   1.0\n",
       "6:152419923 CPCT02060022 ESR1              0/1         1.0   1.0\n",
       "6:152419926 CPCT02160002 ESR1              0/1         1.0   1.0\n",
       "7:140453136 CPCT02010397 BRAF              0/1         1.0   1.0\n",
       "            CPCT02020293 BRAF              0/1         1.0   1.0\n",
       "            CPCT02020360 BRAF              0/1         1.0   1.0\n",
       "            CPCT02060015 BRAF              0/1         1.0   1.0\n",
       "7:55223550  CPCT02010393 EGFR              0/1         1.0   1.0\n",
       "8:128752740 CPCT02070012 MYC               0/1         1.0   1.0\n",
       "9:21971107  CPCT02060019 CDKN2A            0/1         1.0   1.0\n",
       "All                                                   38.0  38.0"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.pivot_table(df[(df.annAllEffects.str.contains(\"missense\"))&(df.normalSGT==\"\")],\\\n",
    "               values='pos', index=['chromPos','patientName','annGene','normalSGT','tumorSGT'], columns=['consensus'], aggfunc='count',margins=True).fillna(\"\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### SUMMARY OF TYPE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>tumorSGT</th>\n",
       "      <th></th>\n",
       "      <th>0/1</th>\n",
       "      <th>1/1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>normalSGT</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <td></td>\n",
       "      <td>67</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0/1</th>\n",
       "      <td>22</td>\n",
       "      <td>1406</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1/1</th>\n",
       "      <td>3</td>\n",
       "      <td></td>\n",
       "      <td>1236.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "tumorSGT        0/1     1/1\n",
       "normalSGT                  \n",
       "                 67     7.0\n",
       "0/1        22  1406     6.0\n",
       "1/1         3        1236.0"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.pivot_table(df, values='pos', index=['normalSGT',], columns=['tumorSGT'], aggfunc='count').fillna(\"\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3> Allelic Depth"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Alllele Freq By Caller\n",
    "for columnName in list(df):\n",
    "    if columnName.endswith('allele'):\n",
    "        ser = df[df[columnName] != ''][columnName[:-6] + 'AF']\n",
    "        ser = ser.sort_values()\n",
    "        #ser[len(ser)] = ser.iloc[-1]\n",
    "        cum_dist = np.linspace(0.,1.,len(ser))\n",
    "        ser_cdf = pd.Series(cum_dist, index=ser,name=columnName[:-6]+\": c=\"+str(ser.count())+\" m=\"+str(round(ser.median(),2)))\n",
    "        ser_cdf.plot(drawstyle='steps',legend=True,title=\" Allelic Frequency by Caller\",figsize=[15,6])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### SCRATCH"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>chromPos</th>\n",
       "      <th>chromFrac</th>\n",
       "      <th>id</th>\n",
       "      <th>ref</th>\n",
       "      <th>vennSegment</th>\n",
       "      <th>numCallers</th>\n",
       "      <th>variantType</th>\n",
       "      <th>variantSubType</th>\n",
       "      <th>filter</th>\n",
       "      <th>bedRegion</th>\n",
       "      <th>inDBSNP</th>\n",
       "      <th>inCOSMIC</th>\n",
       "      <th>annGene</th>\n",
       "      <th>annWorstImpact</th>\n",
       "      <th>annWorstEffect</th>\n",
       "      <th>annAllEffects</th>\n",
       "      <th>consensus</th>\n",
       "      <th>tumorallele</th>\n",
       "      <th>tumorAF</th>\n",
       "      <th>tumorDP</th>\n",
       "      <th>tumorQS</th>\n",
       "      <th>tumorSGT</th>\n",
       "      <th>tumorindelDiff</th>\n",
       "      <th>normalallele</th>\n",
       "      <th>normalAF</th>\n",
       "      <th>normalDP</th>\n",
       "      <th>normalQS</th>\n",
       "      <th>normalSGT</th>\n",
       "      <th>normalindelDiff</th>\n",
       "      <th>patientName</th>\n",
       "      <th>TUMOR_PERCENTAGE</th>\n",
       "      <th>TUMOR_TYPE</th>\n",
       "      <th>PIPELINE_VERSION</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>17</td>\n",
       "      <td>7578115</td>\n",
       "      <td>17:7578115</td>\n",
       "      <td>17.093332</td>\n",
       "      <td>rs1625895</td>\n",
       "      <td>T</td>\n",
       "      <td>test</td>\n",
       "      <td>2</td>\n",
       "      <td>SNP</td>\n",
       "      <td></td>\n",
       "      <td>PASS</td>\n",
       "      <td>TP53</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>TP53</td>\n",
       "      <td>LOW</td>\n",
       "      <td>sequence_feature</td>\n",
       "      <td>sequence_feature|sequence_feature|sequence_fea...</td>\n",
       "      <td>False</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0866142</td>\n",
       "      <td>127</td>\n",
       "      <td>-1</td>\n",
       "      <td>0/1</td>\n",
       "      <td></td>\n",
       "      <td>C</td>\n",
       "      <td>0.435897</td>\n",
       "      <td>39</td>\n",
       "      <td>-1</td>\n",
       "      <td>0/1</td>\n",
       "      <td></td>\n",
       "      <td>CPCT02010393</td>\n",
       "      <td>40</td>\n",
       "      <td>Melanoma</td>\n",
       "      <td>v1.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>17</td>\n",
       "      <td>7578212</td>\n",
       "      <td>17:7578212</td>\n",
       "      <td>17.093333</td>\n",
       "      <td>COSM10654</td>\n",
       "      <td>G</td>\n",
       "      <td>test</td>\n",
       "      <td>1</td>\n",
       "      <td>SNP</td>\n",
       "      <td></td>\n",
       "      <td>PASS</td>\n",
       "      <td>TP53</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>TP53</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>stop_gained|stop_gained|stop_gained|stop_gaine...</td>\n",
       "      <td>False</td>\n",
       "      <td>A</td>\n",
       "      <td>0.838983</td>\n",
       "      <td>118</td>\n",
       "      <td>-1</td>\n",
       "      <td>0/1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>CPCT02010393</td>\n",
       "      <td>40</td>\n",
       "      <td>Melanoma</td>\n",
       "      <td>v1.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>17</td>\n",
       "      <td>7578645</td>\n",
       "      <td>17:7578645</td>\n",
       "      <td>17.093339</td>\n",
       "      <td>rs2909430</td>\n",
       "      <td>C</td>\n",
       "      <td>test</td>\n",
       "      <td>1</td>\n",
       "      <td>SNP</td>\n",
       "      <td></td>\n",
       "      <td>PASS</td>\n",
       "      <td>TP53</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>TP53</td>\n",
       "      <td>LOW</td>\n",
       "      <td>sequence_feature</td>\n",
       "      <td>sequence_feature|sequence_feature|sequence_fea...</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>T</td>\n",
       "      <td>0.548387</td>\n",
       "      <td>31</td>\n",
       "      <td>-1</td>\n",
       "      <td>0/1</td>\n",
       "      <td></td>\n",
       "      <td>CPCT02010393</td>\n",
       "      <td>40</td>\n",
       "      <td>Melanoma</td>\n",
       "      <td>v1.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>17</td>\n",
       "      <td>7579472</td>\n",
       "      <td>17:7579472</td>\n",
       "      <td>17.093349</td>\n",
       "      <td>rs1042522;COSM3766190</td>\n",
       "      <td>G</td>\n",
       "      <td>test</td>\n",
       "      <td>2</td>\n",
       "      <td>SNP</td>\n",
       "      <td></td>\n",
       "      <td>PASS</td>\n",
       "      <td>TP53</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>TP53</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>missense_variant|missense_variant|missense_var...</td>\n",
       "      <td>True</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0915033</td>\n",
       "      <td>153</td>\n",
       "      <td>-1</td>\n",
       "      <td>0/1</td>\n",
       "      <td></td>\n",
       "      <td>C</td>\n",
       "      <td>0.40625</td>\n",
       "      <td>32</td>\n",
       "      <td>-1</td>\n",
       "      <td>0/1</td>\n",
       "      <td></td>\n",
       "      <td>CPCT02010393</td>\n",
       "      <td>40</td>\n",
       "      <td>Melanoma</td>\n",
       "      <td>v1.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>17</td>\n",
       "      <td>37884037</td>\n",
       "      <td>17:37884037</td>\n",
       "      <td>17.466580</td>\n",
       "      <td>rs61552325;COSM5436892</td>\n",
       "      <td>C</td>\n",
       "      <td>test</td>\n",
       "      <td>1</td>\n",
       "      <td>SNP</td>\n",
       "      <td></td>\n",
       "      <td>PASS</td>\n",
       "      <td>ERBB2</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>ERBB2</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>missense_variant|missense_variant|missense_var...</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>G</td>\n",
       "      <td>0.481481</td>\n",
       "      <td>27</td>\n",
       "      <td>-1</td>\n",
       "      <td>0/1</td>\n",
       "      <td></td>\n",
       "      <td>CPCT02010393</td>\n",
       "      <td>40</td>\n",
       "      <td>Melanoma</td>\n",
       "      <td>v1.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>17</td>\n",
       "      <td>38556770</td>\n",
       "      <td>17:38556770</td>\n",
       "      <td>17.474865</td>\n",
       "      <td>rs471692</td>\n",
       "      <td>T</td>\n",
       "      <td>test</td>\n",
       "      <td>2</td>\n",
       "      <td>SNP</td>\n",
       "      <td></td>\n",
       "      <td>PASS</td>\n",
       "      <td>TOP2A</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>TOP2A</td>\n",
       "      <td>LOW</td>\n",
       "      <td>sequence_feature</td>\n",
       "      <td>sequence_feature|intron_variant</td>\n",
       "      <td>False</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>185</td>\n",
       "      <td>-1</td>\n",
       "      <td>1/1</td>\n",
       "      <td></td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>39</td>\n",
       "      <td>-1</td>\n",
       "      <td>1/1</td>\n",
       "      <td></td>\n",
       "      <td>CPCT02010393</td>\n",
       "      <td>40</td>\n",
       "      <td>Melanoma</td>\n",
       "      <td>v1.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>17</td>\n",
       "      <td>41223094</td>\n",
       "      <td>17:41223094</td>\n",
       "      <td>17.507704</td>\n",
       "      <td>rs1799966;COSM3755559</td>\n",
       "      <td>T</td>\n",
       "      <td>test</td>\n",
       "      <td>1</td>\n",
       "      <td>SNP</td>\n",
       "      <td></td>\n",
       "      <td>PASS</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>missense_variant|missense_variant|missense_var...</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>C</td>\n",
       "      <td>0.522727</td>\n",
       "      <td>44</td>\n",
       "      <td>-1</td>\n",
       "      <td>0/1</td>\n",
       "      <td></td>\n",
       "      <td>CPCT02010393</td>\n",
       "      <td>40</td>\n",
       "      <td>Melanoma</td>\n",
       "      <td>v1.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>17</td>\n",
       "      <td>41234470</td>\n",
       "      <td>17:41234470</td>\n",
       "      <td>17.507844</td>\n",
       "      <td>rs1060915;COSM4416375</td>\n",
       "      <td>A</td>\n",
       "      <td>test</td>\n",
       "      <td>1</td>\n",
       "      <td>SNP</td>\n",
       "      <td></td>\n",
       "      <td>PASS</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>LOW</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>synonymous_variant|synonymous_variant|synonymo...</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>G</td>\n",
       "      <td>0.566667</td>\n",
       "      <td>30</td>\n",
       "      <td>-1</td>\n",
       "      <td>0/1</td>\n",
       "      <td></td>\n",
       "      <td>CPCT02010393</td>\n",
       "      <td>40</td>\n",
       "      <td>Melanoma</td>\n",
       "      <td>v1.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>17</td>\n",
       "      <td>41244000</td>\n",
       "      <td>17:41244000</td>\n",
       "      <td>17.507961</td>\n",
       "      <td>rs16942;COSM148277</td>\n",
       "      <td>T</td>\n",
       "      <td>test</td>\n",
       "      <td>2</td>\n",
       "      <td>SNP</td>\n",
       "      <td></td>\n",
       "      <td>PASS</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>missense_variant|missense_variant|missense_var...</td>\n",
       "      <td>True</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0779221</td>\n",
       "      <td>154</td>\n",
       "      <td>-1</td>\n",
       "      <td>0/1</td>\n",
       "      <td></td>\n",
       "      <td>C</td>\n",
       "      <td>0.486486</td>\n",
       "      <td>37</td>\n",
       "      <td>-1</td>\n",
       "      <td>0/1</td>\n",
       "      <td></td>\n",
       "      <td>CPCT02010393</td>\n",
       "      <td>40</td>\n",
       "      <td>Melanoma</td>\n",
       "      <td>v1.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>17</td>\n",
       "      <td>41244435</td>\n",
       "      <td>17:41244435</td>\n",
       "      <td>17.507966</td>\n",
       "      <td>rs16941;COSM3755563</td>\n",
       "      <td>T</td>\n",
       "      <td>test</td>\n",
       "      <td>1</td>\n",
       "      <td>SNP</td>\n",
       "      <td></td>\n",
       "      <td>PASS</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>missense_variant|missense_variant|missense_var...</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>C</td>\n",
       "      <td>0.577778</td>\n",
       "      <td>45</td>\n",
       "      <td>-1</td>\n",
       "      <td>0/1</td>\n",
       "      <td></td>\n",
       "      <td>CPCT02010393</td>\n",
       "      <td>40</td>\n",
       "      <td>Melanoma</td>\n",
       "      <td>v1.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>17</td>\n",
       "      <td>41244936</td>\n",
       "      <td>17:41244936</td>\n",
       "      <td>17.507973</td>\n",
       "      <td>rs799917;COSM148278</td>\n",
       "      <td>G</td>\n",
       "      <td>test</td>\n",
       "      <td>1</td>\n",
       "      <td>SNP</td>\n",
       "      <td></td>\n",
       "      <td>PASS</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>missense_variant|missense_variant|missense_var...</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>A</td>\n",
       "      <td>0.466667</td>\n",
       "      <td>60</td>\n",
       "      <td>-1</td>\n",
       "      <td>0/1</td>\n",
       "      <td></td>\n",
       "      <td>CPCT02010393</td>\n",
       "      <td>40</td>\n",
       "      <td>Melanoma</td>\n",
       "      <td>v1.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>17</td>\n",
       "      <td>41245237</td>\n",
       "      <td>17:41245237</td>\n",
       "      <td>17.507976</td>\n",
       "      <td>rs16940;COSM3755565</td>\n",
       "      <td>A</td>\n",
       "      <td>test</td>\n",
       "      <td>1</td>\n",
       "      <td>SNP</td>\n",
       "      <td></td>\n",
       "      <td>PASS</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>LOW</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>synonymous_variant|synonymous_variant|synonymo...</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>G</td>\n",
       "      <td>0.411765</td>\n",
       "      <td>51</td>\n",
       "      <td>-1</td>\n",
       "      <td>0/1</td>\n",
       "      <td></td>\n",
       "      <td>CPCT02010393</td>\n",
       "      <td>40</td>\n",
       "      <td>Melanoma</td>\n",
       "      <td>v1.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>17</td>\n",
       "      <td>41245466</td>\n",
       "      <td>17:41245466</td>\n",
       "      <td>17.507979</td>\n",
       "      <td>rs1799949;COSM3755567</td>\n",
       "      <td>G</td>\n",
       "      <td>test</td>\n",
       "      <td>1</td>\n",
       "      <td>SNP</td>\n",
       "      <td></td>\n",
       "      <td>PASS</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>sequence_feature</td>\n",
       "      <td>sequence_feature|sequence_feature|synonymous_v...</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>A</td>\n",
       "      <td>0.404255</td>\n",
       "      <td>47</td>\n",
       "      <td>-1</td>\n",
       "      <td>0/1</td>\n",
       "      <td></td>\n",
       "      <td>CPCT02010393</td>\n",
       "      <td>40</td>\n",
       "      <td>Melanoma</td>\n",
       "      <td>v1.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>17</td>\n",
       "      <td>41246481</td>\n",
       "      <td>17:41246481</td>\n",
       "      <td>17.507992</td>\n",
       "      <td>rs1799950;COSM3755568</td>\n",
       "      <td>T</td>\n",
       "      <td>test</td>\n",
       "      <td>2</td>\n",
       "      <td>SNP</td>\n",
       "      <td></td>\n",
       "      <td>PASS</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>missense_variant|missense_variant|missense_var...</td>\n",
       "      <td>True</td>\n",
       "      <td>C</td>\n",
       "      <td>0.94964</td>\n",
       "      <td>139</td>\n",
       "      <td>-1</td>\n",
       "      <td>1/1</td>\n",
       "      <td></td>\n",
       "      <td>C</td>\n",
       "      <td>0.425</td>\n",
       "      <td>40</td>\n",
       "      <td>-1</td>\n",
       "      <td>0/1</td>\n",
       "      <td></td>\n",
       "      <td>CPCT02010393</td>\n",
       "      <td>40</td>\n",
       "      <td>Melanoma</td>\n",
       "      <td>v1.9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom       pos     chromPos  chromFrac                      id ref vennSegment  numCallers variantType variantSubType filter bedRegion inDBSNP inCOSMIC annGene annWorstImpact  \\\n",
       "53    17   7578115   17:7578115  17.093332               rs1625895   T        test           2         SNP                  PASS      TP53    True    False    TP53            LOW   \n",
       "54    17   7578212   17:7578212  17.093333               COSM10654   G        test           1         SNP                  PASS      TP53   False     True    TP53           HIGH   \n",
       "55    17   7578645   17:7578645  17.093339               rs2909430   C        test           1         SNP                  PASS      TP53    True    False    TP53            LOW   \n",
       "56    17   7579472   17:7579472  17.093349   rs1042522;COSM3766190   G        test           2         SNP                  PASS      TP53    True     True    TP53       MODERATE   \n",
       "57    17  37884037  17:37884037  17.466580  rs61552325;COSM5436892   C        test           1         SNP                  PASS     ERBB2    True     True   ERBB2       MODERATE   \n",
       "58    17  38556770  17:38556770  17.474865                rs471692   T        test           2         SNP                  PASS     TOP2A    True    False   TOP2A            LOW   \n",
       "59    17  41223094  17:41223094  17.507704   rs1799966;COSM3755559   T        test           1         SNP                  PASS     BRCA1    True     True   BRCA1       MODERATE   \n",
       "60    17  41234470  17:41234470  17.507844   rs1060915;COSM4416375   A        test           1         SNP                  PASS     BRCA1    True     True   BRCA1            LOW   \n",
       "61    17  41244000  17:41244000  17.507961      rs16942;COSM148277   T        test           2         SNP                  PASS     BRCA1    True     True   BRCA1       MODERATE   \n",
       "62    17  41244435  17:41244435  17.507966     rs16941;COSM3755563   T        test           1         SNP                  PASS     BRCA1    True     True   BRCA1       MODERATE   \n",
       "63    17  41244936  17:41244936  17.507973     rs799917;COSM148278   G        test           1         SNP                  PASS     BRCA1    True     True   BRCA1       MODERATE   \n",
       "64    17  41245237  17:41245237  17.507976     rs16940;COSM3755565   A        test           1         SNP                  PASS     BRCA1    True     True   BRCA1            LOW   \n",
       "65    17  41245466  17:41245466  17.507979   rs1799949;COSM3755567   G        test           1         SNP                  PASS     BRCA1    True     True   BRCA1       MODERATE   \n",
       "66    17  41246481  17:41246481  17.507992   rs1799950;COSM3755568   T        test           2         SNP                  PASS     BRCA1    True     True   BRCA1       MODERATE   \n",
       "\n",
       "        annWorstEffect                                      annAllEffects consensus tumorallele    tumorAF tumorDP tumorQS tumorSGT tumorindelDiff normalallele  normalAF normalDP normalQS normalSGT  \\\n",
       "53    sequence_feature  sequence_feature|sequence_feature|sequence_fea...     False           C  0.0866142     127      -1      0/1                           C  0.435897       39       -1       0/1   \n",
       "54         stop_gained  stop_gained|stop_gained|stop_gained|stop_gaine...     False           A   0.838983     118      -1      0/1                                                                     \n",
       "55    sequence_feature  sequence_feature|sequence_feature|sequence_fea...     False                                                                           T  0.548387       31       -1       0/1   \n",
       "56    missense_variant  missense_variant|missense_variant|missense_var...      True           C  0.0915033     153      -1      0/1                           C   0.40625       32       -1       0/1   \n",
       "57    missense_variant  missense_variant|missense_variant|missense_var...     False                                                                           G  0.481481       27       -1       0/1   \n",
       "58    sequence_feature                    sequence_feature|intron_variant     False           C          1     185      -1      1/1                           C         1       39       -1       1/1   \n",
       "59    missense_variant  missense_variant|missense_variant|missense_var...     False                                                                           C  0.522727       44       -1       0/1   \n",
       "60  synonymous_variant  synonymous_variant|synonymous_variant|synonymo...     False                                                                           G  0.566667       30       -1       0/1   \n",
       "61    missense_variant  missense_variant|missense_variant|missense_var...      True           C  0.0779221     154      -1      0/1                           C  0.486486       37       -1       0/1   \n",
       "62    missense_variant  missense_variant|missense_variant|missense_var...     False                                                                           C  0.577778       45       -1       0/1   \n",
       "63    missense_variant  missense_variant|missense_variant|missense_var...     False                                                                           A  0.466667       60       -1       0/1   \n",
       "64  synonymous_variant  synonymous_variant|synonymous_variant|synonymo...     False                                                                           G  0.411765       51       -1       0/1   \n",
       "65    sequence_feature  sequence_feature|sequence_feature|synonymous_v...     False                                                                           A  0.404255       47       -1       0/1   \n",
       "66    missense_variant  missense_variant|missense_variant|missense_var...      True           C    0.94964     139      -1      1/1                           C     0.425       40       -1       0/1   \n",
       "\n",
       "   normalindelDiff   patientName TUMOR_PERCENTAGE TUMOR_TYPE PIPELINE_VERSION  \n",
       "53                  CPCT02010393               40   Melanoma             v1.9  \n",
       "54                  CPCT02010393               40   Melanoma             v1.9  \n",
       "55                  CPCT02010393               40   Melanoma             v1.9  \n",
       "56                  CPCT02010393               40   Melanoma             v1.9  \n",
       "57                  CPCT02010393               40   Melanoma             v1.9  \n",
       "58                  CPCT02010393               40   Melanoma             v1.9  \n",
       "59                  CPCT02010393               40   Melanoma             v1.9  \n",
       "60                  CPCT02010393               40   Melanoma             v1.9  \n",
       "61                  CPCT02010393               40   Melanoma             v1.9  \n",
       "62                  CPCT02010393               40   Melanoma             v1.9  \n",
       "63                  CPCT02010393               40   Melanoma             v1.9  \n",
       "64                  CPCT02010393               40   Melanoma             v1.9  \n",
       "65                  CPCT02010393               40   Melanoma             v1.9  \n",
       "66                  CPCT02010393               40   Melanoma             v1.9  "
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[(df.patientName=='CPCT02010393')&(df.chrom=='17')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    ""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2.0
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}