---
title: "GRIDSS comparison"
output:
  pdf_document: default
  html_notebook: default
---

# Overview

Hartwig Medical Foundation has been extensively investigating calling 
switching from Manta+BPI to GRIDSS for the detection of structural variants.

The GRIDSS paper can be found at https://genome.cshlp.org/content/early/2017/11/02/gr.222109.117.abstract

A high level overview of the software components of GRIDSS can be found at https://github.com/PapenfussLab/gridss/

Migrating HMF SV calling to GRIDSS will have the following advantages:

## Additional split reads

As part of the GRIDSS pre-processing steps, GRIDSS attempts realignment of all soft-clipped bases for which the aligner has not reported a split read alignment.
Although bwa does report split read alignments, feeding the soft clipped reads back to bwa produces additional alignments which the GRIDSs preprocessor converts to split read alignments.
This results in a slight

## Improved assembly

Manta, like many other SV callers performs targeted breakpoint assembly.
Manta first identifies putative SV location by identifying clusters of read pairs (RP) and split reads (SR).
Targeted local assembly is performed on the candidate SVs and variants are called based on the read and assembly support.
Only a subset of the genome is assembled.

The novel assembly approach taken by GRIDSS is to perform breakend assembly.
GRIDSS assembles all SV-supporting reads across the entire genome with the constraint that assembled reads must have an anchoring read consistent with the assembly.
A soft clipped or split read is only assembled with other soft clipped or split reads that also align to the same genomic loci.
Read pairs with only one read mapped or mapped discordantly are only assembled with reads consistent with the alignment position of the mate and the library fragment size distribution.
As these positional constraints are encoded within the assembly graph itself, we have termed this form of assembly positional de Bruijn graph assembly.
Since fragment will only ever assemble with fragments that have a nearby alignment, we can perform this assembly in a streaming manner - an optimisation not possible with a fully de novo assembly approach.

Once breakend assembly has been performed, the breakpoint partner location must be identified.
To do this GRIDSS realigns the entire contig with bwa then repeatedly realigns the remaining soft-clipped bases until the entire contig is aligned.

### Longer assemblies

In standard de Bruijn graph assembly, simple repeats longer than the k-mer sizes cause loops in the assembly graph.
With a k-mer size of 25, a poly-A sequence of length 27 and 50 both result in the same assembly graph.
Due to this, de Bruijn graph assemblers typically break contigs at these loops.
This causes truncated assembly contigs which are difficult to place.
Since GRIDSS incorporated the expected position into the assembly graph itself the graph is a DAG thus such loops are not present.
Whilst GRIDSS still produces truncated assemblies when the transition from fixed position SR reads to DPs that can be placed in multiple position occurs in k-mer repeat sequence, this is a less common occurrence than any k-mer repeat sequence.

Manta assembled across poly-A sequence in around 15% of CPCT02030318 INS calls. On these same breakpoints, GRIDSS assembled across the poly-A in 60% of cases. For example, at 7:153146820 manta assembled AAAAAACAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA whereas GRIDSS assembled AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAATATAAACAATTTCCTGAATAATAAGTCTTATATAATTATTGAATGATACGTTACTATTTTCCTGTTGGCTTAAATTTTTATTTGTCCGGCGAACCGGAGGGTTTCACGATCCTAATGTCTGTTCAAAAGATTGACTCTTAATCAAAAGAAGAGAGGTTTCGATTCGGTGACAAAC.

## Compound assembly alignment

The most important improvement in switching from manta to GRIDSS is the improvement in the the calling of compound complex events.
When there are multiple nearby linked breakpoints, breakpoints signals will overlap.
Take the example of a rearrangement of DNA segments A, B, C, and D which have been rearranged to be adjacent.
If segments B and C are small, say 200bp each, then there will be read pairs corrected aligned to A/B, A/C, A/D, B/C, B/D, and C/D that support the breakpoint.
Ideally, this results in variant calls A-B, B-C, C-D and `IMPRECISE` variant calls supported only by read pairs for A-C, A-D, and B-D.
In practice, when B and C are short, manta/BPI filters out many of the B-C calls due to the lack of RP support.

In contrast, GRIDSS was explicitly designed to handle the complex rearrangments present in tumour neochromosomes.
Since GRIDSS contigs are repeatedly realigned, each contig can support multiple breakpoints.
A 600bp contig assembled at A, will align to A, B, C, and D since it spans all these events.
When calling B-C, not only are the reads aligned to B and C considered to support the breakpoint, but all assembly contigs spanning the B-C breakpoint are included.
This results in a strong signal for the B-C and C-D breakpoints.

Looking at the raw assembly contig alignments for CPCT02030318, of the alignments that align back to the originating location and at least one other location, 8% have multiple alignments, with 22% of those having at least 3 alignment locations in addition to the originating location.

## Breakend calling

In addition to your typical SV breakpoints, GRIDSS reports single breakends.
Single breakends are locations in which there is evidence a structural variant occurs but the partner breakend cannot be unambiguously identified.
Typically these are events that have one side in repetitive sequence, but can also be the result in incorporation of non-reference sequence such as retroviral insertions events.
Such events are able to pinpoint changes in copy number that would not otherwise be able to the placed to base-pair accuracy.
Knowning the existence of such events enables a more accurate reconstruction of the somatic genome.
As these events required a different data model, they have not yet been incorporated into the Hartwig pipeline.

----

## Outstanding concerns

### File handles

GRIDSS will temporarily leak file handles until the next full GC.
In certain situations, these unused file handles can accumulate until the file handle ulimit is reached and GRIDSS is killed byt he OS.
GRIDSS using htsjdk to handle SAM/BAM IO which memory maps BAM index files.
Due to security concerns, Java does not provide an API to explicitly close memory mapping.
As a result, used file handles are not closed until a GC disposes of the underlying channel.

The workaround for this behaviour is to increase the file handle ulimit on the OS (ulimit -n).





